README.md
<!-- -*- mode:markdown; coding:utf-8; -*- -->

# MPRTest : A demonstration of the Minkowski Portal Refinement algorithm #

Demonstrates intersection tests with the Minkowski Portal Refinement
(MPR) algorithm, introduced by Gary Snethen.

*   Last Update: Jan 08, 2022
*   Since: Nov 16, 2008

Example:

*   Cone-Cylinder (not colliding)

       <img src="https://raw.githubusercontent.com/vaiorabbit/MPRTest/master/doc/Result00.png" width="350">

*   Cone-Cylinder (detected colliding position and penetration)

       <img src="https://raw.githubusercontent.com/vaiorabbit/MPRTest/master/doc/Result01.png" width="350">


## Prerequisites ##

*   Ruby Interpreter <http://www.ruby-lang.org/>
    *   Version : 2.4.0 or higher
    *   RubyInstaller For Windows :
        *   http://rubyinstaller.org

*   GLFW
  *   Copy GLFW DLL (glfw3.dll or libglfw.dylib) here.
  *   GLFW (https://www.glfw.org).

*   opengl-bindings
    *   Available via RubyGems.

            $ gem install opengl-bindings2

    *   Ref.:
        *   https://rubygems.org/gems/opengl-bindings
        *   https://github.com/vaiorabbit/ruby-opengl

*   rmath3d
    *   Available via RubyGems

            $ gem install rmath3d_plain

    *   Ref.:
        *    https://rubygems.org/gems/rmath3d
        *    https://github.com/vaiorabbit/rmath3d

## Usage ##

    X:\> ruby Application.rb

*   The two objects are rendered with transparency when the intersection
    test (`MPRAlgorithm::get_contact()`) returns `true`;


## Operation ##

*   Q or Esc     : quit.
*   Space        : reset the objects' setting.
*   WASD         : move red object.
*   Z            : change type of shape0 (red)
*   IJKL         : move blue object.
*   M            : change type of shape1 (blue)
*   Mouse L/R    : move eye position.


## Reference ##

*   Game Programming Gems 7 <http://www.charlesriver.com/Books/BookDetail.aspx?productID=373>
    *   See "2.5 XenoCollide: Complex Collision Made Simple" for algorithm details.

*   XenoCollide <http://xenocollide.com>
    *   Support site maintained by Gary Snethen, the author of the GPG7 article.

*   mpr2d <http://code.google.com/p/mpr2d/>
    *   A 2D implementation of the MPR algorithm by Mason Green.


## License ##

All source codes are available under the terms of the zlib/libpng license
(see LICENSE.txt).

-------------------------------------------------------------------------------

# MPRTest : Minkowski Portal Refinement アルゴリズムのデモプログラム #

Gary Snethen 氏による Minkowski Portal Refinement (MPR) アルゴリズムのデモプログラムです。


## 必要なもの ##

*   Ruby <http://www.ruby-lang.org/>
    *   バージョン : 2.4.0 以降
    *   RubyInstaller For Windows :
        *   http://rubyinstaller.org

*   GLFW
  *   GLFWのDLL (glfw3.dll もしくは libglfw.dylib) をここにコピーしてください。
  *   GLFW (https://www.glfw.org).

*   opengl-bindings
    *   RubyGems でインスールできます。

            $ gem install opengl-bindings2

    *   参考:
        *   https://rubygems.org/gems/opengl-bindings
        *   https://github.com/vaiorabbit/ruby-opengl

*   rmath3d
    *   RubyGems でインスールできます。

            $ gem install rmath3d_plain

    *   Ref.:
        *    https://rubygems.org/gems/rmath3d
        *    https://github.com/vaiorabbit/rmath3d


## 使い方 ##

    X:\> ruby Application.rb

*   交差判定 (`MPRAlgorithm::get_contact()`) が true となる場合に、2つの箱が半透明で描画されます。


## Operation ##

*   Q または Esc         : プログラムの終了
*   Space                : オブジェクトの設定をリセット
*   WASD                 : 赤いオブジェクトの移動
*   Z                    : 赤いオブジェクトの形状変更
*   IJKL                 : 青いオブジェクトの移動
*   M                    : 青いオブジェクトの形状変更
*   マウスでの L/R       : 視点の移動


## 参考文献 ##

*   Game Programming Gems 7 <http://www.charlesriver.com/Books/BookDetail.aspx?productID=373>
    *   "2.5 XenoCollide: Complex Collision Made Simple" にアルゴリズムの詳細が掲載されています。

*   XenoCollide <http://xenocollide.com>
    *   Gary Snethen 氏（上記GPG7の著者）によるサポートサイトです。

*   mpr2d <http://code.google.com/p/mpr2d/>
    *   Mason Green 氏による 2D版のデモプログラムです。


## License ##

ソースコードは zlib/libpng ライセンスの条件下で利用可能です。


ChangeLog
2022-01-08  vaiorabbit  <http://twitter.com/vaiorabbit>

	* Update: Now uses opengl-bindings2 gem ( https://rubygems.org/gems/opengl-bindings ) and GLFW ( https://glfw.org ).

2013-09-22  vaiorabbit  <http://twitter.com/vaiorabbit>

	* Update: Now uses opengl-bindings gem ( https://rubygems.org/gems/opengl-bindings ).
	* Update: Removed rmath3d_plain.rb. rmath3d is now available as a gem package ( https://rubygems.org/gems/rmath3d ).

2013-07-20  vaiorabbit  <http://twitter.com/vaiorabbit>

	* Update: Old contact information is updated to the current one.
	* Math functions replaced with new 'rmath3d'.

2009-12-26  LT140

	* Update: Old contact information is updated to the current one.

2009-06-14  LT140

	* Application.rb : New shape `MPRAlgorithm::Triangle'.
	* Shape.rb (Triangle): Added.
	* Draw.rb (self.triangle): Added.
	
2008-11-24  LT140

	* Shape.rb (Cylinder,Cone,Sphere): Added.

2008-11-16  LT140

	* Announce: First release.


UnitTest/test.rb
require_relative '../MPRAlgorithm'
require_relative '../Shape'

require 'minitest/autorun'
require_relative 'test_MPRAlgorithm.rb'


UnitTest/test_MPRAlgorithm.rb
require 'rmath3d/rmath3d'
include RMath3D

class TC_MPRAlgorithm < MiniTest::Test
  def setup
  end

  def teardown
  end

  def test_intersect_boxes
    box0 = MPRAlgorithm::Box.new
    box1 = MPRAlgorithm::Box.new

    box0.center.setElements( 0.0, 0.0, 0.0 )
    box0.half.setElements( 0.5, 0.5, 0.5 )
    box0.orientation = RQuat.new( 0.0,0.0,0.0, 1.0 )

    box1.center.setElements( 1.0, 0.0, 0.0 )
    box1.half.setElements( 0.5, 0.5, 0.5 )
    box1.orientation = RQuat.new.rotationAxis( RVec3.new(0,0,1), Math::PI/4 )

    result = MPRAlgorithm.intersect( box0, box1 )
    assert( true == result )
  end
end


Draw.rb
require 'opengl'
require 'rmath3d/rmath3d_plain'
require_relative 'Geometry'

module Draw

  include RMath3D

  def self.initialize
    init_floor()
    init_box()
    init_cylinder()
    init_cone()
    init_sphere()
    init_point()
    init_arrow()
  end

  def self.finalize
    displaylists = [@@dl_box, @@dl_cylinder, @@dl_cone, @@dl_sphere, @@dl_point, @@dl_arrow]
    displaylists.each do |dl|
      GL.DeleteLists(dl, 1)
    end
  end

  def self.init_floor()
    pos_max = 10.0
    pos_min = -pos_max

    nlines = 10

    stride = (pos_max - pos_min) / nlines

    pos = Array.new(3 * 2*(nlines+1), 0.0)

    for i in 0..nlines do
      pos[6*i+0] = pos_min
      pos[6*i+1] = 0.0
      pos[6*i+2] = pos_min+stride*i

      pos[6*i+3] = pos_max
      pos[6*i+4] = 0.0
      pos[6*i+5] = pos_min+stride*i
    end

    for i in 0..nlines do
      j = i + nlines+1

      pos[6*j+0] = pos_min+stride*i
      pos[6*j+1] = 0.0
      pos[6*j+2] = pos_min

      pos[6*j+3] = pos_min+stride*i
      pos[6*j+4] = 0.0
      pos[6*j+5] = pos_max
    end

    @@packed_pos = pos.pack("d*")
    @@packed_pos_indices = pos.length/3
    pos = nil
  end

  def self.floor()
    GL.PushAttrib(GL::ALL_ATTRIB_BITS)

    GL.LineWidth(1.0)
    GL.Disable(GL::LIGHTING)
    GL.ShadeModel(GL::FLAT)
    GL.Color4fv([ 0.2, 0.2, 0.8, 1.0 ].pack('F4'))

    GL.EnableClientState(GL::VERTEX_ARRAY)
    GL.VertexPointer(3, GL::DOUBLE, 0, @@packed_pos)
    GL.DrawArrays(GL::LINES, 0, @@packed_pos_indices)
    GL.DisableClientState(GL::VERTEX_ARRAY)

    GL.PopAttrib()
  end

  def self.init_box
    @@dl_box = GL.GenLists(1)
    GL.NewList(@@dl_box, GL::COMPILE)
    geomSolidCube(1)
    GL.EndList()
  end

  def self.box(pos, quat, half)
    GL.PushMatrix()

    mtx = RMtx4.new.rotationQuaternion(quat)
    mtx.e03 = pos.x
    mtx.e13 = pos.y
    mtx.e23 = pos.z
    GL.MultMatrixf(mtx.to_a.pack('F16'))

    GL.PushMatrix()
    GL.Scalef(2.0*half.x, 2.0*half.y, 2.0*half.z)
    GL.CallList(@@dl_box)
    GL.PopMatrix()

    GL.PopMatrix()
  end

  def self.init_cylinder
    @@dl_cylinder = GL.GenLists(1)
    GL.NewList(@@dl_cylinder, GL::COMPILE)

    n = 18

    GL.Begin(GL::QUADS)
    n.times do |i|
      z0 = Math.cos(360*i/n * Math::PI/180)
      x0 = Math.sin(360*i/n * Math::PI/180)
      z1 = Math.cos(360*(i+1)/n * Math::PI/180)
      x1 = Math.sin(360*(i+1)/n * Math::PI/180)
      GL.Normal3f(x0, 0.0, z0)
      GL.Vertex3f(x0,  0.5, z0)
      GL.Normal3f(x0, 0.0, z0)
      GL.Vertex3f(x0, -0.5, z0)
      GL.Normal3f(x1, 0.0, z1)
      GL.Vertex3f(x1, -0.5, z1)
      GL.Normal3f(x1, 0.0, z1)
      GL.Vertex3f(x1,  0.5, z1)
    end
    GL.End()

    GL.Begin(GL::TRIANGLE_FAN)
    GL.Normal3f(0.0, 1.0, 0.0)
    GL.Vertex3f(0.0, 0.5, 0.0)
    n.times do |i|
      z0 = Math.cos(360*i/n * Math::PI/180)
      x0 = Math.sin(360*i/n * Math::PI/180)
      z1 = Math.cos(360*(i+1)/n * Math::PI/180)
      x1 = Math.sin(360*(i+1)/n * Math::PI/180)
      GL.Normal3f(0.0, 1.0, 0.0)
      GL.Vertex3f(x0,  0.5, z0)
      GL.Normal3f(0.0, 1.0, 0.0)
      GL.Vertex3f(x1,  0.5, z1)
    end
    GL.End()

    GL.Begin(GL::TRIANGLE_FAN)
    GL.Normal3f(0.0, -1.0, 0.0)
    GL.Vertex3f(0.0, -0.5, 0.0)
    n.downto(1) do |i|
      z0 = Math.cos(360*i/n * Math::PI/180)
      x0 = Math.sin(360*i/n * Math::PI/180)
      z1 = Math.cos(360*(i-1)/n * Math::PI/180)
      x1 = Math.sin(360*(i-1)/n * Math::PI/180)
      GL.Normal3f(0.0, -1.0, 0.0)
      GL.Vertex3f(x0,  -0.5, z0)
      GL.Normal3f(0.0, -1.0, 0.0)
      GL.Vertex3f(x1,  -0.5, z1)
    end
    GL.End()

    GL.EndList()
  end

  def self.cylinder(pos, quat, radius, half_height)
    GL.PushMatrix()

    mtx = RMtx4.new.rotationQuaternion(quat)
    mtx.e03 = pos.x
    mtx.e13 = pos.y
    mtx.e23 = pos.z
    GL.MultMatrixf(mtx.to_a.pack('F16'))

    GL.Scalef(radius, 2.0*half_height, radius)
    GL.CallList(@@dl_cylinder)
    GL.PopMatrix()
  end


  def self.init_cone
    @@dl_cone = GL.GenLists(1)
    GL.NewList(@@dl_cone, GL::COMPILE)

    n = 18

    GL.Begin(GL::TRIANGLE_FAN)
    GL.Normal3f(0.0, 1.0, 0.0)
    GL.Vertex3f(0.0, 1.0, 0.0)
    n.times do |i|
      z0 = Math.cos(360*i/n * Math::PI/180)
      x0 = Math.sin(360*i/n * Math::PI/180)
      z1 = Math.cos(360*(i+1)/n * Math::PI/180)
      x1 = Math.sin(360*(i+1)/n * Math::PI/180)
      GL.Normal3f(x0, 0.0, z0)
      GL.Vertex3f(x0, 0.0, z0)
      GL.Normal3f(x1, 0.0, z1)
      GL.Vertex3f(x1, 0.0, z1)
    end
    GL.End()

    GL.Begin(GL::TRIANGLE_FAN)
    GL.Normal3f(0.0, -1.0, 0.0)
    GL.Vertex3f(0.0,  0.0, 0.0)
    n.downto(1) do |i|
      z0 = Math.cos(360*i/n * Math::PI/180)
      x0 = Math.sin(360*i/n * Math::PI/180)
      z1 = Math.cos(360*(i-1)/n * Math::PI/180)
      x1 = Math.sin(360*(i-1)/n * Math::PI/180)
      GL.Normal3f(0.0, -1.0, 0.0)
      GL.Vertex3f(x0,   0.0, z0)
      GL.Normal3f(0.0, -1.0, 0.0)
      GL.Vertex3f(x1,   0.0, z1)
    end
    GL.End()

    GL.EndList()
  end


  def self.cone(pos, quat, radius, height)
    GL.PushMatrix()

    mtx = RMtx4.new.rotationQuaternion(quat)
    mtx.e03 = pos.x
    mtx.e13 = pos.y
    mtx.e23 = pos.z
    GL.MultMatrixf(mtx.to_a.pack('F16'))

    GL.Scalef(radius, height, radius)
    GL.CallList(@@dl_cone)
    GL.PopMatrix()
  end

  def self.init_sphere
    @@dl_sphere = GL.GenLists(1)
    GL.NewList(@@dl_sphere, GL::COMPILE)
    geomSolidSphere(1.0, 16, 16)
    GL.EndList()
  end

  def self.sphere(pos, radius)
    GL.PushMatrix()
    GL.Translatef(pos.x, pos.y, pos.z)
    GL.Scalef(radius, radius, radius)
    GL.CallList(@@dl_sphere)
    GL.PopMatrix()
  end

  def self.init_point
    @@dl_point = GL.GenLists(1)
    GL.NewList(@@dl_point, GL::COMPILE)
    GL.Color4f(1,1,0,1)
    geomSolidSphere(1.0, 16, 16)
    GL.EndList()
  end

  def self.point(pos)
    radius = 0.05
    GL.PushMatrix()
    GL.Translatef(pos.x, pos.y, pos.z)
    GL.Scalef(radius, radius, radius)
    GL.CallList(@@dl_point)
    GL.PopMatrix()
  end

  def self.init_arrow
    @@dl_arrow = GL.GenLists(1)
    GL.NewList(@@dl_arrow, GL::COMPILE)
    GL.Color4f(0,0.8,0,1)
    geomSolidCone(0.1, 0.5, 8, 8)
    GL.EndList()
  end

  def self.arrow(pos, dir, depth, basis)
    line_translation = pos + depth * dir

    GL.Begin(GL::LINES)
    GL.Color4f(1,1,0,1)
    GL.Vertex3f(pos.x, pos.y, pos.z)
    GL.Color4f(0,0.8,0,1)
    GL.Vertex3f(line_translation.x, line_translation.y, line_translation.z)
    GL.End()

    mtxRotY = RMtx3.new.rotationY(90.0 * Math::PI/180.0)
    mtx = basis * mtxRotY
    mtx4x4 = RMtx4.new(mtx[0,0], mtx[0,1], mtx[0,2], 0.0,
                        mtx[1,0], mtx[1,1], mtx[1,2], 0.0,
                        mtx[2,0], mtx[2,1], mtx[2,2], 0.0,
                        0.0,      0.0,      0.0,      1.0)
    GL.PushMatrix()
    GL.Translatef(line_translation.x, line_translation.y, line_translation.z)
    GL.MultMatrixf(mtx4x4.to_a.pack('F16'))
    GL.CallList(@@dl_arrow)
    GL.PopMatrix()
  end

  def self.triangle(*pos)
#     raise ArgumentError if pos.length != 3
#     raise ArgumentError if pos[0].class != RVec3 || pos[1].class != RVec3 || pos[2].class != RVec3

    GL.Begin(GL::TRIANGLES)
    normal = RVec3.cross(pos[2]-pos[0], pos[1]-pos[0]).normalize!()
    GL.Normal3fv(normal.to_a.pack('F3'))
    GL.Vertex3fv(pos[0].to_a.pack('F3'))
    GL.Vertex3fv(pos[1].to_a.pack('F3'))
    GL.Vertex3fv(pos[2].to_a.pack('F3'))

    normal = RVec3.cross(pos[1]-pos[0], pos[2]-pos[0]).normalize!()
    GL.Normal3fv(normal.to_a.pack('F3'))
    GL.Vertex3fv(pos[0].to_a.pack('F3'))
    GL.Vertex3fv(pos[2].to_a.pack('F3'))
    GL.Vertex3fv(pos[1].to_a.pack('F3'))
    GL.End();
  end

end


=begin
MPRTest : A demonstration program of Minkowski Portal Refinement.
Copyright (c) 2008-2022 vaiorabbit <http://twitter.com/vaiorabbit>

This software is provided 'as-is', without any express or implied
warranty. In no event will the authors be held liable for any damages
arising from the use of this software.

Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

    1. The origin of this software must not be misrepresented; you must not
    claim that you wrote the original software. If you use this software
    in a product, an acknowledgment in the product documentation would be
    appreciated but is not required.

    2. Altered source versions must be plainly marked as such, and must not be
    misrepresented as being the original software.

    3. This notice may not be removed or altered from any source
    distribution.
=end


Camera.rb
require 'opengl'
require 'rmath3d/rmath3d_plain'

class Camera
  include RMath3D

  attr_reader :mtxView

  def initialize
    @mouse_state = 0
    @prev_x = 0
    @prev_y = 0
    @phi = Math::PI/2.0
    @theta = Math::PI/3.0
    @radius = 20.0

    @position = RVec3.new(@radius * Math.sin(@theta) * Math.cos(@phi), @radius * Math.cos(@theta), @radius * Math.sin(@theta) * Math.sin(@phi))
    @at       = RVec3.new(0, 0, 0)
    @up       = RVec3.new(0, 1, 0)

    @mtxView = RMtx4.new.lookAtRH(@position, @at, @up)
  end

  def set_mouse_state(button, state, x, y)
    if button == GLFW::MOUSE_BUTTON_LEFT
      if state == GLFW::PRESS
        @mouse_state |= 1
      else # GLFW::RELEASE
        @mouse_state &= ~1
      end
    end
    if button == GLFW::MOUSE_BUTTON_RIGHT
      if state == GLFW::PRESS
        @mouse_state |= 2
      else # GLFW::RELEASE
        @mouse_state &= ~2
      end
    end
    if button == GLFW::MOUSE_BUTTON_MIDDLE
      if state == GLFW::PRESS
        @mouse_state |= 4
      else # GLFW::RELEASE
        @mouse_state &= ~4
      end
    end
    @prev_x = x
    @prev_y = y
  end

  def update_from_mouse_motion(x, y)
    if @mouse_state != 0
      dx = (x - @prev_x).to_f
      dy = (y - @prev_y).to_f

      if @mouse_state == 1 # Left
        scale = 0.5
        @phi += scale * dx * Math::PI/180
        @theta -= scale * dy * Math::PI/180
      elsif @mouse_state == 2 # Right
        scale = 0.5
        @radius -= scale * dy
      end

      wrap_params()
      @position.x = @radius * Math.sin(@theta) * Math.cos(@phi)
      @position.z = @radius * Math.sin(@theta) * Math.sin(@phi)
      @position.y = @radius * Math.cos(@theta)

      @renew_view = true
      @renew_pvinv = true
    end

    @prev_x = x
    @prev_y = y
  end

  def wrap_params
    @phi -= 2*Math::PI if @phi > 2*Math::PI
    @phi += 2*Math::PI if @phi < -2*Math::PI

    @theta = Math::PI/2 if @theta > Math::PI/2
    @theta = RMath3D::TOLERANCE if @theta < RMath3D::TOLERANCE

    @radius = 30.0 if @radius > 30.0
    @radius = 10.0 if @radius < 10.0
  end

  def load_camera_matrix
    GL.MatrixMode(GL::MODELVIEW)
    GL.LoadIdentity()

    GL.MultMatrixf(@mtxView.lookAtRH(@position, @at, @up).to_a.pack('F16'))
  end

end


=begin
MPRTest : A demonstration program of Minkowski Portal Refinement.
Copyright (c) 2008-2022 vaiorabbit <http://twitter.com/vaiorabbit>

This software is provided 'as-is', without any express or implied
warranty. In no event will the authors be held liable for any damages
arising from the use of this software.

Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

    1. The origin of this software must not be misrepresented; you must not
    claim that you wrote the original software. If you use this software
    in a product, an acknowledgment in the product documentation would be
    appreciated but is not required.

    2. Altered source versions must be plainly marked as such, and must not be
    misrepresented as being the original software.

    3. This notice may not be removed or altered from any source
    distribution.
=end


Application.rb
require 'opengl'
require 'glfw'
require 'rmath3d/rmath3d_plain'

require_relative 'Draw'
require_relative 'Camera'
require_relative 'MPRAlgorithm'
require_relative 'Shape'

def glfw_library_path()
  case GL.get_platform
  when :OPENGL_PLATFORM_WINDOWS
    Dir.pwd + '/glfw3.dll'
  when :OPENGL_PLATFORM_MACOSX
    './libglfw.dylib'
  when :OPENGL_PLATFORM_LINUX
    '/usr/lib/x86_64-linux-gnu/libglfw.so' # not tested
  else
    raise RuntimeError, "Unsupported platform."
  end
end

# module Fiddle
#   class Closure
#     class BlockCaller
#       def self.finalize(id)
#         puts "Object #{id} dying at #{Time.new}"
#       end
#     end
#   end
# end

class Application

  def draw
    @camera.load_camera_matrix()

    GL.Clear(GL::COLOR_BUFFER_BIT | GL::DEPTH_BUFFER_BIT)
    GL.PushAttrib(GL::ALL_ATTRIB_BITS)

    GL.Lightfv(GL::LIGHT0, GL::POSITION, [10,10,10,1].pack('F4'))
    GL.Lightfv(GL::LIGHT0, GL::DIFFUSE,  [1,1,1,1].pack('F4'))
    GL.Lightfv(GL::LIGHT0, GL::SPECULAR, [1,1,1,1].pack('F4'))
    GL.Lightfv(GL::LIGHT0, GL::AMBIENT,  [0.2,0.2,0.2,1].pack('F4'))

    GL.Lightfv(GL::LIGHT1, GL::POSITION, [-10,-10,-10,1].pack('F4'))
    GL.Lightfv(GL::LIGHT1, GL::DIFFUSE,  [1,1,1,1].pack('F4'))
    GL.Lightfv(GL::LIGHT1, GL::SPECULAR, [1,1,1,1].pack('F4'))
    GL.Lightfv(GL::LIGHT1, GL::AMBIENT,  [0.2,0.2,0.2,1].pack('F4'))

    Draw.floor()

    # Render two shapes. Transparency is enabled when they overlap.

    GL.DepthMask(GL::FALSE) if @overlap # For rendering with transparency.

    GL.Materialfv(GL::FRONT, GL::SPECULAR, [0,0,0,1].pack('F4'))
    GL.Materialfv(GL::FRONT, GL::AMBIENT, [0.2,0.2,0.2,1].pack('F4'))
    GL.Materialf(GL::FRONT, GL::SHININESS, 0.0)

    shapes = [@shape0, @shape1].sort_by { |s| s.center.transformCoord(@camera.mtxView).z }
    shapes.each do |s|
      GL.Materialfv(GL::FRONT, GL::DIFFUSE, [s.rgb, @overlap ? 0.75 : 1.0].flatten!.pack('F4'))
      s.draw
    end

    if @overlap.class == MPRAlgorithm::ContactInfo
      @overlap.update_basis
      GL.DepthMask(GL::TRUE)
      GL.Disable(GL::DEPTH_TEST)
      GL.Disable(GL::LIGHTING)
      Draw.point(@overlap.position)
      Draw.arrow(@overlap.position, @overlap.normal, @overlap.penetration, @overlap.basis)
    end

    GL.PopAttrib()

    GLFW.SwapBuffers(@window)
    GLFW.PollEvents()
  end
  private :draw

  # :GLFWkeyfun
  def key(window, key, scancode, action, mods)
    if key == GLFW::KEY_ESCAPE && action == GLFW::PRESS
      GLFW.SetWindowShouldClose(window, 1)
    end

    if action == GLFW::PRESS
      case key
      when GLFW::KEY_SPACE
        # 'Space' : Reset posture of all shapes.
        @shape0.reset
        @shape1.reset
        # @overlap = MPRAlgorithm.intersect(@shape0, @shape1)
        @overlap = MPRAlgorithm.get_contact(@shape0, @shape1)
        return
      when GLFW::KEY_Z
        # 'z' : Change type of shape0
        center = RMath3D::RVec3.new(@shape0.center)
        @shape0 = new_shape(@rgb_red)
        @shape0.center = center
        @overlap = MPRAlgorithm.get_contact(@shape0, @shape1)
        return
      when GLFW::KEY_M
        # 'm' : Change type of shape1
        center = RMath3D::RVec3.new(@shape1.center)
        @shape1 = new_shape(@rgb_blue)
        @shape1.center = center
        @overlap = MPRAlgorithm.get_contact(@shape0, @shape1)
        return
      end
    end

    shape0move = RMath3D::RVec3.new
    shape1move = RMath3D::RVec3.new

    case key
    # Move red shape.
    when GLFW::KEY_A ; shape0move.x -= 0.125
    when GLFW::KEY_S ; shape0move.z += 0.125
    when GLFW::KEY_W ; shape0move.z -= 0.125
    when GLFW::KEY_D ; shape0move.x += 0.125

    # Move blue shape.
    when GLFW::KEY_J ; shape1move.x -= 0.125
    when GLFW::KEY_K ; shape1move.z += 0.125
    when GLFW::KEY_I ; shape1move.z -= 0.125
    when GLFW::KEY_L ; shape1move.x += 0.125
    end

    # Transform the movement in view space into world space.
    mtxViewInv = @camera.mtxView.getInverse
    shape0move.transformNormal!(mtxViewInv)
    shape1move.transformNormal!(mtxViewInv)
    shape0move.y = 0
    shape1move.y = 0

    @shape0.center.add!(shape0move)
    @shape1.center.add!(shape1move)

    # @overlap = MPRAlgorithm.intersect(@shape0, @shape1)
    @overlap = MPRAlgorithm.get_contact(@shape0, @shape1)
  end
  private :key

  # :GLFWwindowsizefun
  def reshape(window, width, height)
    @window_width  = width
    @window_height = height

    width_buf = ' ' * 8
    height_buf = ' ' * 8
    GLFW.GetFramebufferSize(@window, width_buf, height_buf)

    GL.Viewport(0, 0, width_buf.unpack1('L'), height_buf.unpack1('L'))
    GL.MatrixMode(GL::PROJECTION)
    GL.LoadIdentity()
    GL.MultMatrixf(RMath3D::RMtx4.new.perspectiveFovRH(30.0*Math::PI/180.0, width.to_f/height.to_f, 0.1, 1000.0).to_a.pack('F16'))
  end
  private :reshape

  # :GLFWmousebuttonfun
  def mouse(window, button, action, mods)
    x_buf = ' ' * 8
    y_buf = ' ' * 8
    GLFW.GetCursorPos(window, x_buf, y_buf)
    x = x_buf.unpack1('D')
    y = y_buf.unpack1('D')

    @camera.set_mouse_state(button, action, x, y)
  end
  private :mouse

  # :GLFWcursorposfun
  def motion(window, x, y)
    @camera.update_from_mouse_motion(x, y)
  end
  private :motion

  def new_shape(rgb)
    s = @shape_classes.rotate![0].new
    s.reset
    s.rgb = rgb
    return s
  end

  def initialize

    GLFW.load_lib(glfw_library_path())
    GLFW.Init()

    @window = GLFW.CreateWindow(640, 360, "MPRAlgorithm::intersect Test", nil, nil)

    return if @window.null?

    GLFW.MakeContextCurrent(@window)

    GL.load_lib()

    @key_callback = GLFW::create_callback(:GLFWkeyfun, method(:key))
    @size_callback = GLFW::create_callback(:GLFWwindowsizefun, method(:reshape))
    @mousebutton_callback = GLFW::create_callback(:GLFWmousebuttonfun, method(:mouse))
    @cursorpos_callback = GLFW::create_callback(:GLFWcursorposfun, method(:motion))
    GLFW.SetKeyCallback(@window, @key_callback)
    GLFW.SetWindowSizeCallback(@window, @size_callback)
    GLFW.SetMouseButtonCallback(@window, @mousebutton_callback)
    GLFW.SetCursorPosCallback(@window, @cursorpos_callback)

    @size_callback.call(@window, 640, 360)

    # ObjectSpace.define_finalizer(@key_callback,
    #                              @key_callback.class.method(:finalize).to_proc)

    # Common render states
    GL.Enable(GL::NORMALIZE)
    GL.Enable(GL::CULL_FACE)
    GL.Enable(GL::DEPTH_TEST)
    GL.DepthFunc(GL::LEQUAL)
    GL.DepthMask(GL::TRUE)
    GL.Enable(GL::BLEND)
    GL.BlendFunc(GL::SRC_ALPHA, GL::ONE_MINUS_SRC_ALPHA)
    GL.Enable(GL::LIGHTING)
    GL.Enable(GL::LIGHT0)
    GL.Enable(GL::LIGHT1)
    GL.ClearColor(0.9, 0.9, 0.95, 1.0)

    Draw.initialize

    @camera = Camera.new

    @shape_classes = [MPRAlgorithm::Box,
                      MPRAlgorithm::Cylinder,
                      MPRAlgorithm::Cone,
                      MPRAlgorithm::Sphere,
                      MPRAlgorithm::Triangle
                     ]

    @rgb_red  = [1.0,0.1,0.1]
    @rgb_blue = [0.1,0.1,1.0]

    @shape0 = new_shape(@rgb_red)
    @shape1 = new_shape(@rgb_blue)
    # @overlap = MPRAlgorithm.intersect(@shape0, @shape1)
    @overlap = MPRAlgorithm.get_contact(@shape0, @shape1)
  end

  def run
    while GLFW.WindowShouldClose(@window) == 0
      draw
    end
  end

  def finalize
    Draw.finalize
  end
end


if __FILE__ == $PROGRAM_NAME
  begin
    app = Application.new
    app.run
  ensure
    app&.finalize
  end
end

=begin
MPRTest : A demonstration program of Minkowski Portal Refinement.
Copyright (c) 2008-2022 vaiorabbit <http://twitter.com/vaiorabbit>

This software is provided 'as-is', without any express or implied
warranty. In no event will the authors be held liable for any damages
arising from the use of this software.

Permission is granted to anyone to use this software for any purpose,
including commercial applications, and to alter it and redistribute it
freely, subject to the following restrictions:

    1. The origin of this software must not be misrepresented; you must not
    claim that you wrote the original software. If you use this software
    in a product, an acknowledgment in the product documentation would be
    appreciated but is not required.

    2. Altered source versions must be plainly marked as such, and must not be
    misrepresented as being the original software.

    3. This notice may not be removed or altered from any source
    distribution.
=end
